SOLR-16687: Add a SolrClassLoader to SolrZkClient

laminelam opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16687

 
# Description

A Solr class loader is needed in SolrZKClient to instantiate classes part of _modules_.

This is part of a bigger PR to add support of AWS Secret Manager to store Zookeeper credentials.  
 
# Tests

Local integration tests.

Do we need a unit test here?

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

gerlowskija reviewed (commented):
This looks 👍  so far.  Given a bit more time for other folks to review and some automated tests, I'd be happy to merge!

HoustonPutman commented:
I think the classloader needs to be used in both `createZkCredentialsToAddAutomatically` and `createZkACLProvider` as well.

laminelam commented:
> This looks 👍 so far. Given a bit more time for other folks to review and some automated tests, I'd be happy to merge!

Hi @gerlowskija 
Added a test case to test SCL based instantiation, will add more tests (SC vs DCL scenarios) as part of the main PR to add support of AWS Secret Manager.

I have also changed the behavior in case of an Exception is raised when instantiating _ZkACLProvider_ or _ZkCredentialsInjector_, it should fail-fast rather then ignoring it and go with the default _ZkACLProvider_/_ZkCredentialsInjector_ which sets znodes permissions  to OPEN_ACL_UNSAFE.

laminelam commented:
> I think the classloader needs to be used in both `createZkCredentialsToAddAutomatically` and `createZkACLProvider` as well.

Hi @HoustonPutman 

I think a solrClassLoader is not needed for _ZkACLProvider_ and _ZkCredentialsProvider_  except if we add in the future custom implementations part of another contrib/module, but it doesn't hurt to add it now. I have added it to both methods.

gerlowskija commented on solr/solrj-zookeeper/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java:
```diff
@@ -281,4 +282,42 @@ public void testCheckInterrupted() {
+        .withSolrClassLoader(solrResourceLoader)
+        .build()
+        .close(); // no more tests needed. We only test class instantiation
+    clearSystemProperties();
```
[Q] Do you know whether these system properties actually need cleared?  Solr tests all run with a JUnit '@Rule' (called "SystemPropertiesRestoreRule") that is supposed to clear them in between each test case.

If that's not working, I can file a separate ticket to fix that.  But if it is, we should remove redundant sysprop removal here.

laminelam replied:
Thanks @gerlowskija , didn't know that _SystemPropertiesRestoreRule_ was part of _SolrTestCase_
Actually I followed the existing style, as I have seen this logic used in other classes.
Removed it.
Updated changes.txt file as well.
Thank you for your time on this

gerlowskija commented on solr/solrj-zookeeper/src/test/org/apache/solr/common/cloud/SolrZkClientTest.java:
```diff
@@ -281,4 +282,42 @@ public void testCheckInterrupted() {
+    clearSystemProperties();
+  }
+
+  private static void setupSystemProperties() {
```
[0] Small nit: this might benefit from a name that describes what sysprops you're setting up. e.g. `enableCustomCredentialsProvider` or something similar.

But if you like that name as-is, that's also fine.

laminelam replied:
done :)

gerlowskija reviewed (approved):
This LGTM from a testing perspective; it's not ideal but I've wracked my brain trying to come up with a more comprehensive way to do this, without much luck.

It'd be good to give Houston another few days to chime in though, since he had other concerns to speak to.

HoustonPutman commented:
My concerns have been addressed!
